src/components/organisms/Badges/Badges.tsx

```diff
@@ -0,0 +1,165 @@
+
+  const visitHours = useMemo(() => {
+    if (!visits) return 0;
+
+    const visitSeconds = visits
+      .filter((v) => v.id === currentVenue.id || relatedVenueIds.includes(v.id))
+      .reduce((acc, v) => acc + v.timeSpent, 0);
+    return Math.floor(visitSeconds / (60 * 60));
+  }, [visits, currentVenue, relatedVenueIds]);
+
+  // Only show visits to related venues
+  const relevantVisits = visits?.filter(
+    (visit) =>
+      currentVenue.id === visit.id || relatedVenueIds.includes(visit.id)
+  );
+
+  const badges = useMemo(() => {
+    if (!relevantVisits || !venues) return [];
+    return relevantVisits
+      .map((visit) => {
+        const { venue, room } = findVenueAndRoomByName(visit.id, venues);
+        if (
+          !venue ||
+          (currentVenue.id !== venue.id && !relatedVenueIds.includes(venue.id))
+        )
+          return undefined;
+
+        if (room) {
+          return {
+            venue,
+            room,
+            image: room.image_url,
+            label: room.title,
+          };
+        }
+
+        return {
+          venue,
+          image: venue.mapIconImageUrl,
+          label: venue.name,
+        };
+      })
+      .filter((b) => b !== undefined);
+  }, [relevantVisits, venues, currentVenue, relatedVenueIds]);
+
+  if (!relevantVisits) {
+    return <>Visit venues to collect badges!</>;
+  }
+
+  return (
+    <>
+      <div className="visits">
+        <div className="visit-item">
+          <span className="visit-item__value">
+            {visitHours > 1 ? `${visitHours}` : "< 1"} hrs
+          </span>
+          <span className="visit-item__label">Time spent in Sparkle</span>
+        </div>
+        <div className="separator"></div>
+        <div className="visit-item">
+          <span className="visit-item__value">
+            {relevantVisits?.length ?? 0}
+          </span>
+          <span className="visit-item__label">Venues visited</span>
+        </div>
+      </div>
+      <div className="badges-container">
+        <div className="badges-title">{badges.length} Badges</div>
+        <ul className="badge-list">
+          {badges.map(
+            (b) =>
+              b && (
+                <li className="badge-list-item" key={b.label}>
+                  <Link to={getLocationLink(b.venue, b.room)}>
+                    <img
+                      className="badge-list-item-image"
+                      src={b.image}
+                      alt={`${b.label} badge`}
+                    />
+                  </Link>
+                </li>
+              )
+            // label missing on hover - similar to playa
+          )}
+        </ul>
+      </div>
+    </>
+  );
+};
+
+const findVenueAndRoomByName = (
+  nameOrRoomTitle: string,
+  venues: Array<WithId<AnyVenue>>
+) => {
+  const venue = venues.find(
+    (v) =>
+      v.name === nameOrRoomTitle ||
+      (isCampVenue(v) && v.rooms.find((r) => r.title === nameOrRoomTitle))
+  );
```

Comment:
Will this need to use the 'venue with rooms' thing instead of 'camp venue'? We've hit a few bugs where things didn't work as expected because they were using this directly.

Also, does this bigger function make sense to move to a `utils/` or similar for reusability?

Reply:
resolved in https://github.com/sparkletown/sparkle/pull/895/commits/3e72c18e563d7fbd06de2c6c287223f3a54ec7be#

Reply:
[Future] extract `findVenueAndRoomByName` function from here into a `utils/*` file for reusability.